### PR TITLE
add early stopping

### PR DESCRIPTION
**Pull request by sho-asn**
@superleesa 
I'm gonna debug the early stopping on the remote desktop. Once I could successfully get the result. Ill ask you to merge.

### CODE REVIEW

File: train/sft/toenet_base/train.py [hunk 1]
```diff
@@ -128,8 +131,17 @@ def train_loop(
...
+                early_stopping(val_loss)
+                if early_stopping.early_stop:
+                    print("Early stopping")
+                    break 
+
             global_step_counter += 1
 
+        if early_stopping.early_stop:
+            print("Early stopping")
+            break
```
Comment (superleesa) [inline train/sft/toenet_base/train.py]: why are you printing "early stopping" in both line 136 and 142?

File: train/uft/train.py [hunk 2]
```diff
@@ -267,6 +269,15 @@ def train_loop(
             torch.cuda.empty_cache()
             global_step_counter += 1
 
+            early_stopping(vaL_denoiser_loss)
```
Comment (superleesa) [inline train/uft/train.py]: for UFT, you also need to look at val_discriminator_loss. if both val_discriminator_loss and val_denoiser_loss are not getting better in "n" consecutive epochs, then you can stop? (i think you need to change the __call__ method of the EarlyStopping class to require two parameters, not one)

**Review by superleesa — CHANGES_REQUESTED** [review]

File: train/sft/toenet_base/train.py (hunk 1 above)
Comment (sho-asn) [reply train/sft/toenet_base/train.py]: @superleesa 
Early stopping should occur if the validation loss decreases in **n consecutive epochs**? I'm wondering if I should write it in the nested loop or in the loop (one for loop)

File: train/uft/train.py (hunk 2 above)
Comment (sho-asn) [reply train/uft/train.py]: @superleesa 
What should I set for default "n"? 3 or 5?

Comment (superleesa) [reply train/uft/train.py]: I guess for now 5 is fine

File: train/sft/toenet_base/train.py (hunk 1 above)
Comment (superleesa) [reply train/sft/toenet_base/train.py]: I think the problem is that we are calculating the validation loss too often. 

Can you change the code so that we only calculate the validation loss at the end of each epoch? Then we can determine to early stop after the validation loss computation.

Comment (sho-asn) [reply train/sft/toenet_base/train.py]: @superleesa 
okay, you mean we only need to calculate the validation loss one time every epoch right?

Comment (superleesa) [reply train/sft/toenet_base/train.py]: yup

File: train/uft/train.py (hunk 2 above)
Comment (sho-asn) [reply train/uft/train.py]: @superleesa 
if we only calculate the validation loss once every epoch, isn't 3 better? Cuz the default epoch is 20 for sft and 10 for uft.

Comment (superleesa) [reply train/uft/train.py]: In terms of patience, a higher value is more conservative than lower values. I simply thought that if we set the patience too small, the training can stop "too early" even though it still has a potential to be a better model. We can first try with 5 and then reduce to smaller value if 5 was too much.

Comment (sho-asn) [reply train/uft/train.py]: @superleesa 
okay, i got it

**Review by superleesa — APPROVED** [review]
